fusor/models.py

```diff
@@ -609,12 +566,18 @@ def enforce_element_quantities(cls, values):
             raise ValueError(uq_gene_msg)
         return values
 
-    @root_validator(skip_on_failure=True)
+    @model_validator(mode="after")
     def structural_elements_ends(cls, values):
         """Ensure start/end elements are of legal types and have fields
         required by their position.
         """
-        elements = values.get("structural_elements", [])
+        if isinstance(values, dict):
+            try:
+                elements = values.get("structural_elements", [])
+            except KeyError:
+                raise TypeError
+        else:
+            elements = values.structural_elements
```

Comment:
If using after validator 

```suggestion
        elements = values.structural_elements
```